apps/src/codegen.js

```diff
@@ -48,6 +48,31 @@ exports.evalWith = function (code, options, legacy) {
   }
 };
 
+exports.evalWithEvents = function (apis, events, generator) {
+  let interpreter, currentCallback;
+  const hooks = {};
+  const code = Object.keys(events).map(event => {
+    hooks[event] = () => {
+      currentCallback(event);
+      interpreter.run();
+    };
+    return `function ${event} () {
+      ${generator(events[event])}
+    }
+    `;
```

Comment:
will this approach end up adding unwanted extra whitespace to the produced code?